plugins/plugin-git/che-plugin-git-ext-git/src/main/java/org/eclipse/che/ide/ext/git/client/GitChangesHandler.java

```diff
@@ -0,0 +1,140 @@
+ *******************************************************************************/
+package org.eclipse.che.ide.ext.git.client;
+
+import org.eclipse.che.api.core.jsonrpc.commons.RequestHandlerConfigurator;
+import org.eclipse.che.api.git.shared.Status;
+import org.eclipse.che.api.project.shared.dto.event.GitChangeEventDto;
+import org.eclipse.che.ide.api.editor.EditorAgent;
+import org.eclipse.che.ide.api.notification.NotificationManager;
+import org.eclipse.che.ide.api.parts.EditorMultiPartStack;
+import org.eclipse.che.ide.api.parts.EditorTab;
+import org.eclipse.che.ide.api.resources.File;
+import org.eclipse.che.ide.api.resources.Resource;
+import org.eclipse.che.ide.api.vcs.VcsStatus;
+import org.eclipse.che.ide.part.explorer.project.ProjectExplorerPresenter;
+import org.eclipse.che.ide.resource.Path;
+import org.eclipse.che.ide.resources.tree.FileNode;
+import org.eclipse.che.ide.resources.tree.ResourceNode;
+import org.eclipse.che.ide.ui.smartTree.Tree;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import static org.eclipse.che.ide.api.vcs.VcsStatus.ADDED;
+import static org.eclipse.che.ide.api.vcs.VcsStatus.MODIFIED;
+import static org.eclipse.che.ide.api.vcs.VcsStatus.NOT_MODIFIED;
+import static org.eclipse.che.ide.api.vcs.VcsStatus.UNTRACKED;
+
+/**
+ * Receives and handel's git changes notifications caught by server side VFS file watching system.
+ *
+ * @author Igor Vinokur
+ */
+@Singleton
+public class GitChangesHandler {
+
+    private final Provider<EditorAgent>              editorAgentProvider;
+    private final Provider<ProjectExplorerPresenter> projectExplorerPresenterProvider;
+    private final Provider<EditorMultiPartStack>     multiPartStackProvider;
+
+    @Inject
+    public GitChangesHandler(RequestHandlerConfigurator configurator,
+                             Provider<EditorAgent> editorAgentProvider,
+                             Provider<ProjectExplorerPresenter> projectExplorerPresenterProvider,
+                             Provider<EditorMultiPartStack> multiPartStackProvider) {
+        this.editorAgentProvider = editorAgentProvider;
+        this.projectExplorerPresenterProvider = projectExplorerPresenterProvider;
+        this.multiPartStackProvider = multiPartStackProvider;
+
+        configureHandler(configurator);
+    }
+
+    private void configureHandler(RequestHandlerConfigurator configurator) {
+        configurator.newConfiguration()
+                    .methodName("event:git-change")
+                    .paramsAsDto(GitChangeEventDto.class)
+                    .noResult()
+                    .withBiConsumer(this::apply);
+
+        configurator.newConfiguration()
+                    .methodName("event:git-index")
+                    .paramsAsDto(Status.class)
+                    .noResult()
+                    .withBiConsumer(this::apply);
+    }
+
+    public void apply(String endpointId, GitChangeEventDto dto) {
```

Comment:
Thank you for the description, if @vparfonov is okay with it, let's leave it in current state.